tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/CardinalityConstRule.scala

```diff
@@ -0,0 +1,87 @@
+package at.forsyte.apalache.tla.bmcmt.rules
+
+import at.forsyte.apalache.tla.bmcmt._
+import at.forsyte.apalache.tla.bmcmt.rules.aux.CherryPick
+import at.forsyte.apalache.tla.lir.convenience._
+import at.forsyte.apalache.tla.lir.oper.{BmcOper, TlaArithOper, TlaFiniteSetOper}
+import at.forsyte.apalache.tla.lir.values.TlaInt
+import at.forsyte.apalache.tla.lir.{OperEx, TlaEx, ValEx}
+import at.forsyte.apalache.tla.bmcmt.implicitConversions._
+import at.forsyte.apalache.tla.bmcmt.types.BoolT
+
+
+/**
+  * Optimization for Cardinality(S) >= k, where k is constant. See [docs/smt/Cardinality.md].
+  *
+  * @author Igor Konnov
+  */
+class CardinalityConstRule(rewriter: SymbStateRewriter) extends RewritingRule {
+  private val pickRule = new CherryPick(rewriter)
+
+  override def isApplicable(state: SymbState): Boolean = {
+    state.ex match {
+      case OperEx(BmcOper.constCard,
+      OperEx(TlaArithOper.ge, OperEx(TlaFiniteSetOper.cardinality, _), ValEx(TlaInt(_)))) =>
+        true
+
+      case _ =>
+        false
+    }
+  }
+
+  override def apply(state: SymbState): SymbState = {
+    state.ex match {
+      case OperEx(BmcOper.constCard,
+      OperEx(TlaArithOper.ge, OperEx(TlaFiniteSetOper.cardinality, setEx), ValEx(TlaInt(thresholdBigInt)))) =>
+        val threshold = thresholdBigInt.toInt
+        var nextState = rewriter.rewriteUntilDone(state.setRex(setEx).setTheory(CellTheory()))
+        val setCell = nextState.asCell
+        val elems = nextState.arena.getHas(setCell)
+        if (threshold <= 0) {
+          nextState.setRex(tla.bool(true))
+        } else if (elems.size < threshold) {
+          nextState.setRex(tla.bool(false))
+        } else {
+          mkWitnesses(nextState, setCell, elems, threshold)
+        }
+
+      case _ =>
+        throw new RewriterException("%s is not applicable".format(getClass.getSimpleName), state.ex)
+    }
+  }
+
+  private def mkWitnesses(state: SymbState, set: ArenaCell, elems: Seq[ArenaCell], threshold: Int): SymbState = {
+    def solverAssert = rewriter.solverContext.assertGroundExpr(_)
+
+    var nextState = state
+    nextState = nextState.updateArena(_.appendCell(BoolT()))
+    val emptyPred = nextState.arena.topCell
+    solverAssert(tla.eql(emptyPred, tla.and(elems.map(e => tla.not(tla.in(e.toNameEx, set.toNameEx))) :_*)))
+
+    // pick `threshold` cells that will act as witnesses
+    def pick(i: Int): ArenaCell = {
+      nextState = pickRule.pick(set, nextState, emptyPred.toNameEx)
+      nextState.asCell
+    }
+
+    val witnesses = 1.to(threshold).map(pick).toList
+
+    // create an inequality constraint
+    def mkEq(pair: (ArenaCell, ArenaCell)): TlaEx = {
+      nextState = rewriter.lazyEq.cacheOneEqConstraint(nextState, pair._1, pair._2)
+      tla.not(tla.eql(pair._1, pair._2))
+    }
+
+    // create the inequality predicate
+    val cons = (witnesses cross witnesses).filter(p => p._1.id < p._2.id).map(mkEq)
```

Comment:
I am not sure about the efficiency, but at least it would make both the code, as well as the SMT encoding smaller, if we used the `distinct` predicate, available in smt-lib and in Z3 solver context (`mkDistinct`)

Reply:
Nice! I did not know about `distinct`. There might be other places in the code, where `distinct` could be used. Just to note, `mkEq` also creates equalities constraints, if needed.

Reply:
I have implemented this test with `(distinct ...)`. However, our benchmarks have too small thresholds to see the effect.